server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -699,7 +715,7 @@ public void onFailure(Exception e) {
 
                                 @Override
                                 protected void doRun() throws Exception {
-                                    doExecute(task, bulkRequest, actionListener);
+                                    doDispatchedExecute(task, bulkRequest, actionListener);
```

Comment:
We avoid double accounting because on callback we call `doInternalExecute` vs. `doExecute`. `doInternalExecute` does not account.

> Also, in case where we forward the bulk request (see forwardIngestRequest), we can probably stop accounting for the bytes of the original bulk request as soon as the request is sent.

I'm not sure I understand the logic here. How is this different than the node acting as coordinator node for a normal indexing request? Once we forward a request to a primary we are done until we have to forward the response.

Reply:
> I'm not sure I understand the logic here. How is this different than the node acting as coordinator node for a normal indexing request? Once we forward a request to a primary we are done until we have to forward the response.

In this case where we forward the bulk request to another ingest node, we're actually not holding onto the request bytes anymore (as there is no retry mechanism at this level), so we could in theory stop accounting for it. I'm not sure yet though if this optimization is relevant at this point, or whether we want to make this more resilient and introduce retries at this level as well.

Reply:
Sure, but we are stilling holding the bytes. Particularly if we are doing the "unsafe" thing at the http level.

We could not mark the bytes at the `TransportBulkAction` level if you prefer. For example, we still would mark the coordinating bytes at the `TransportShardBulkAction` level. Or we could just mark bytes for actual ingest work. And then unmark when we are done (they will be marked again once we become the coordinating node).

Reply:
I prefer the extra marking (i.e. err on the side of overaccounting instead of underaccounting). If we don't mark at the `TransportBulkAction` level, then we won't account for requests that are waiting on auto-index creations or local ingest pipeline transformations. Let's keep this as is for now.